API v3 Fix basilist scroll giving logic when joining party

KristianTashkov opened the pull request:
While testing the guild code I noticed that `group.memberCount` is not set to 1 when we create a group, so the first member to join (after the creator) becomes the leader, so I fixed it in this commit. I can remove it and add it in a separate PR if you want.

UserID: bb8db09b-5822-4608-bba3-1486964b8537

KristianTashkov committed:
Fix basilist scroll giving logic when joining party

crookedneighbor commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -34,6 +34,7 @@ api.createGroup = {
 
     let group = new Group(Group.sanitize(req.body)); // TODO validate empty req.body
     group.leader = user._id;
+    group.memberCount = 1;
```
In my opinion, this is something that should happen in the `new Group` call.

crookedneighbor replied:
Maybe not. I'm of two minds about it.

KristianTashkov replied:
I have no strong opinion about it, the place where you add the first user and the leader seemed the most appropriate.

paglias replied:
We could add 1 as the default value for memberCount in the model

crookedneighbor commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -257,11 +260,13 @@ api.joinGroup = {
+      }
+
+      Q.all(promises)
+      .then(() => {
```
Let's put this `.then` On the same line as `Q.all`

crookedneighbor commented:
I think the controller code looks sound. I have some nitpicky thoughts about test structure that I will try to post tomorrow.

KristianTashkov commented:
I fixed the code style issue and added expects for the `group.memberCount = 1` in the appropriate tests.

paglias commented:
@KristianTashkov this looks good, the only thing is that tests for invite route should be in a different file called `POST-groups_groupId_join`

TheHollidayInn commented on test/api/v3/integration/groups/POST-groups_groupId_join.js:
```diff
@@ -0,0 +1,63 @@
+      })
+      .then((guild) => {
+        return generateUser({ 'invitations.guilds': [{ id: guild._id}]}).then((generatedUser) => {
+          return generatedUser.post('/groups/' + guild._id + '/join');
```
Is this okay to generate an invitation this way? Can we do the same for simply adding users to groups? This is the style I was talking about for skipping routes and only testing the route we need to call. @paglias @crookedneighbor

paglias replied:
in this case it was done because we don't have the invite route ready yet, in general I don't know. To me it doesn't look like calling real routes makes things more difficult, rather it should help with routes that have a lot of side effects that are more difficult to replicate. But I'll leave the decision to @crookedneighbor

KristianTashkov replied:
While writing it I changed my opinion on what's better like 5 times. On one hand calling the API is better, but sometimes you need to set a lot of other stuff to make an API call work, that will just set a variable for you. It slows down the unit test, but also makes it resilient to change in the implementation of said API route. If we don't care about unit tests running time and the API call can be executed without a lot of hassle, I think it's the better choice. If it's easier to mock the API route, than to set up the API call to actually run, maybe mocking is better.

TheHollidayInn replied:
Yea, I agree. And, @crookedneighbor, I were discussing it. I couldn't think of the word, but Mocking was what I was thinking of. There is a difference between integration tests and unit tests though. Unit tests you definitely want to mock all this.

crookedneighbor replied:
I [wrote a helper](https://github.com/HabitRPG/habitrpg/blob/02f4a841463c87159041c2e66b2faec60caa140d/test/helpers/api-integration.helper.js#L132-L212) (for v2 routes) to generate a group with members and invitees.

Once the invite routes are done, that should be used instead. Until them, I'd be fine with adding a todo comment to impliment it.

KristianTashkov commented:
@crookedneighbor  Rewrote using `await` syntax :heart:

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_groupId_join.js:
```diff
@@ -0,0 +1,51 @@
+  });
+
+  context('Guilds', () => {
+    it('inviting to guild doesn\'t give basilist to inviter', async () => {
```
This should read more naturally in English: `it inviting to guild doesn't give basilist to inviter`

Perhaps:

```
it('does not give basilist quest to inviter when joining a guild', async () => {
```

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_groupId_join.js:
```diff
@@ -0,0 +1,51 @@
+describe('POST /group/:groupId/join', () => {
+  let user;
+
+  beforeEach(() => {
```
You can make this:

```
beforeEach(async () => {
  user = await generateUser();
});
```

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_groupId_join.js:
```diff
@@ -0,0 +1,51 @@
+    });
+  });
+
+  context('Guilds', () => {
```
I'd make this context something like: `Accepting an invitation to a guild`

KristianTashkov replied:
Same as party context.

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_groupId_join.js:
```diff
@@ -0,0 +1,51 @@
+    });
+  });
+
+  context('Parties', () => {
```
I'd make this context something like: `Accepting an invitation to a party`

KristianTashkov replied:
I plan to add more tests here for the join route once this PR is merged.

crookedneighbor replied:
I think it should be done now.

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_groupId_join.js:
```diff
@@ -0,0 +1,51 @@
+
+  context('Guilds', () => {
+    it('inviting to guild doesn\'t give basilist to inviter', async () => {
+      let groupName = 'Test Guild';
```
I'd move the guild creation into a beforeEach block in this context.

KristianTashkov replied:
Same answer as before, when I add the new tests the plan is to use more beforeEach blocks for common logic.

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_groupId_join.js:
```diff
@@ -0,0 +1,51 @@
+      let groupName = "Test Party";
+      let groupType = "party";
+
+      user = await generateUser({'items.quests.basilist': 0 });
```
I don't think this user should be instantiated with `'items.quests.basilist': 0`. The test should cover that it handles users who do not have the scroll at all yet.

In addition, there should be an additional test that has a user that is instantiated with `'items.quests.basilist': 5` (the 5 is arbitrary, it should just be greater than 0) to show that it increments the value

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_groupId_join.js:
```diff
@@ -0,0 +1,51 @@
+
+  context('Parties', () => {
+    it('gives basilist quest item to the inviter', async () => {
+      let groupName = "Test Party";
```
The party created should be put into a beforeEach in this context.

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_groupId_join.js:
```diff
@@ -0,0 +1,51 @@
+      await invitedUser.post('/groups/' + party._id + '/join');
+      
+      let changedUser = await user.get('/user');
+      expect(changedUser.items.quests.basilist).to.equal(1);
```
For things like this, I'm unsure of which syntax we want to use. We could do it like you have it, or we could do it like this:

``` js
await expect(user.get('/user')).to.eventually.have.deep.property('items.quest.basilist', 1);
```

Does anyone have thoughts?

crookedneighbor commented on website/src/controllers/api-v3/groups.js:
```diff
@@ -257,11 +259,12 @@ api.joinGroup = {
-        user.save(),
-        User.update({_id: user.invitations.party.inviter}, {$inc: {'items.quests.basilist': 1}}).exec(), // Reward inviter
-      ]).then(() => {
+      let promises = [group.save(), user.save()];
```
Can we add a new line here?

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_groupId_join.js:
```diff
@@ -0,0 +1,51 @@
+      });
+
+      let invitedUser = await generateUser({ 'invitations.guilds': [{ id: guild._id}]});
+      await invitedUser.post('/groups/' + guild._id + '/join');
```
Where possible, we should be using template strings:

```
`/groups/${guild._id}/join`
```

KristianTashkov commented:
@crookedneighbor I fixed most of your notes. When I added more tests to handle both cases where the user has scrolls and doesn't then we are changing the user that creates the group, hence his creation and the group's can't be in a `beforeEach`. For future tests we can add a context with a valid creator where we don't care about basilist, but for now I just don't see a way to include these two cases and still have the creation in a `beforeEach`.

crookedneighbor commented on test/api/v3/integration/groups/POST-groups_groupId_join.js:
```diff
@@ -0,0 +1,84 @@
+    it('does not give basilist quest to inviter when joining a guild', async () => {
+      user = await generateUser({balance: 1})
+      group = await user.post('/groups', {
+        name: groupName,
```
I don't think these need to be variables. There's no need to assert on them later. Why not just:

```
{
  name: 'Test Guild',
  type: 'guild',
}
```

crookedneighbor commented:
I've been toying with adding an update method to the API user class for scenarios like this.

crookedneighbor commented:
Updated it to use beforeEach and `user.update`

Closed with https://github.com/HabitRPG/habitrpg/commit/7025bbfa3244dc1da9477513b6db68f4e436bf17

Remind me of your Habitica UUID?
